controllers/virtualmachine/virtualmachine_controller.go

```diff
@@ -156,6 +156,26 @@ func classToVMMapperFn(ctx *pkgctx.ControllerManagerContext, c client.Client, is
 	}
 }
 
+func upgradeSchema(ctx *pkgctx.VirtualMachineContext) {
+	// If empty, this VM was created before v1alpha3 added the spec.biosUUID field.
+	if ctx.VM.Spec.BiosUUID == "" && ctx.VM.Status.BiosUUID != "" {
+		ctx.VM.Spec.BiosUUID = ctx.VM.Status.BiosUUID
+		ctx.Logger.Info("Upgrade VirtualMachine spec",
+			"biosUUID", ctx.VM.Spec.BiosUUID)
+	}
+
+	// If empty, this VM was created before v1alpha3 added the instanceID field
+	if bs := ctx.VM.Spec.Bootstrap; bs != nil {
+		if ci := bs.CloudInit; ci != nil {
+			if ci.InstanceID == "" {
+				ci.InstanceID = string(ctx.VM.UID)
+				ctx.Logger.Info("Upgrade VirtualMachine spec",
+					"bootstrap.cloudInit.instanceID", ci.InstanceID)
```

Comment:
This is missing the cloud-init instance ID annotation case. Should we replicate the following logic here?

https://github.com/vmware-tanzu/vm-operator/blob/main/pkg/providers/vsphere/vmlifecycle/bootstrap_cloudinit.go#L53-L103

Reply:
Right, yes. Pushed a change so we can use the same logic in both places.